src/classes/UTIL_Finder.cls

```diff
@@ -0,0 +1,360 @@
+*/
+/**
+* @author Salesforce.org
+* @date 2018
+* @group Utilities
+* @description The SObject Finder use to query, filter, sort and limit number of records by using SOQL or SOSL
+*/
+public with sharing class UTIL_Finder implements UTIL_QueryApi.Finder {
+    /** @description Search exception class */
+    public class SoslException extends Exception {}
+
+    /** @description Message to indicate the Search is constructed incorrectly when sObjectType is not provided */
+    @TestVIsible private static final String SOBJECT_TYPE_REQUIRED = 'sObjectType is required for the search';
+    /** @description Message to indicate the Search is constructed incorrectly when no value is specified in FIND clause */
+    @TestVIsible private static final String SEARCH_CRITERIA_REQUIRED = 'Search criteria is required for the search';
+    /** @description Message to indicate the Search is constructed incorrectly when no sObject field is specified */
+    @TestVIsible private static final String FIELDS_REQUIRED = 'At least one field required to be returned from the search';
+
+    /** @description The sObjectType to query/search */
+    private sObjectType sObjType;
+    /** @description Fields to return from the query/search */
+    @TestVisible private Set<String> selectFields {
+        get {
+            if (selectFields == null) {
+                selectFields = new Set<String>{ 'Id' };
+            }
+            return selectFields;
+        }
+        set;
+    }
+    /** @description The WHERE clause */
+    @TestVisible private UTIL_Where criteria;
+    /** @description The ORDER BY segment */
+    private UTIL_OrderBy orderBy;
+    /** @description The LIMIT segment */
+    private Integer limitRows;
+
+    /*******************************************************************************************************************
+    * @description THe constructor for the provided sObjectType
+    */
+    public UTIL_Finder(sObjectType sObjType) {
+        this.sObjType = sObjType;
+        this.orderBy = new UTIL_OrderBy();
+        this.criteria = new UTIL_Where();
+    }
+
+    /*********************************************************************************************************
+    * @description Sets SELECT fields in the SOQL or RETURNING fields in the SOSL
+    * @param sObjFields Fields
+    * @return UTIL_Finder The object instance
+    */
+    public UTIL_Finder withSelectFields(Set<Schema.sObjectField> sObjFields) {
+        for (Schema.sObjectField sObjField : sObjFields) {
+            selectFields.add(String.valueOf(sObjField));
+        }
+
+        return this;
+    }
+
+    /*********************************************************************************************************
+    * @description Sets SELECT fields in the SOQL or RETURNING fields in the SOSL
+    * @param fieldSet The Field Set fields
+    * @return UTIL_Finder The object instance
+    */
+    public UTIL_Finder withSelectFields(List<FieldSetMember> fieldSet) {
```

Comment:
I really like how all of these methods return the `UTIL_Finder` instance so methods can be stacked into a single call.